src/components/FormStep.js

```diff
@@ -264,47 +341,82 @@ const FormStep = ({form, submission, onLogicChecked, onStepSubmitted, onLogout})
-  // things using immer.produce to deal with immutable state inside at some point.
+  /**
+   * Closes the save confirmation modal.
+   *
+   * During evaluation, the following actions may be dispatched:
+   *
+   *   - `TOGGLE_FORM_SAVE_MODAL` Immediately, resulting in `isFormSaveModalOpen=false`, closing
+   *     modal.
+   */
+  const closeFormStepSaveModal = () => {
+    dispatch({type: 'TOGGLE_FORM_SAVE_MODAL', payload: {open: false}});
+  };
+
+  /**
+   * Event loops and async programming are fun!
+   *
+   * UI inputs are wonky if end-users perform input while evaluating logic checks that operate on
+   * (slightly) stale form data. Evaluating the ref value once before something doing IO and using
+   * that value _after_ the IO event completed can lead to de-sync. This utility ensures you always
+   * have an up-to-date view of the form data.
+   *
+   * Currently it's a simple wrapper around the form ref, but we may do more advanced things using
+   * immer.produce to deal with immutable state inside at some point.
+   *
+   * @return {Object|null}
+   */
   const getCurrentFormData = () => {
     const submissionData = formRef.current?.formio?.submission?.data;
     return submissionData ? {...submissionData} : null;
   };
 
+  /**
+   * Check whether a current logic using the _check_Logic endpoint should be cancelled.
+   * @throws {AbortedLogicCheck} Throws an AbortedLogicCheck if the current logic check should be
+   *   aborted
+   * @param {*} signal
+   */
   const checkAbortedLogicCheck = signal => {
     const shouldAbortCurrentCheck = signal.aborted;
-    if (!shouldAbortCurrentCheck) return;
+    if (!shouldAbortCurrentCheck) {
+      return;
+    }
     // throw custom error object to exit current callback forcibly
     throw new AbortedLogicCheck('Aborted logic check');
   };
 
   /**
-   * Evaluate the server side logic
-   * @param  {AbortController} controller     AbortController used to abort XHR requests and/or result processing
-   *   because of new user input.
-   * @param  {Boolean} canSubmitState The original canSubmit state at the time of scheduling the logic check.
-   * @param  {Boolean} forceEvaluation Force re-evaluation of the logic check even if the data hasn't changed (#2488).
-   * @return {Void}                No return, dispatches reducer actions leading to state updates.
+   * Evaluates the server side logic.
+   *
+   * During evaluation, the following actions may be dispatched:
+   *
+   *   - `LOGIC_CHECK_INTERRUPTED` When there's no point/change in state to be expected by the logic
+   *       check
+   *   - `BLOCK_SUBMISSION` When the the logic is checking and the form should be disabled for
+   *       submission.
+   *   - `LOGIC_CHECK_DONE` When the logic check is done and the form should be enabled for
+   *       submission.
+   *   - `LOGIC_CHECK_INTERRUPTED` When an error occurred during the logic check.
```

Comment:
you are tying the dispatch action states with the state mutations that happen in the reducer here, this will definitely go out of sync if/when we update internal state aspects in the reducers.

The action types are intended to be sufficiently descriptive to have a broad idea of what is happening/being affected, the detailed documentation belongs in the reducer.

These descriptions also don't capture all nuances, e.g. `LOGIC_CHECK_INTERRUPTED` is not because some error happened, it may just have been part of the debounce mechanism. You probably mean that a technical exception was raised to communicate that, but that's just an implementation detail that is not relevant for the functional side of things.

Reply:
Fro me "action types are intended to be sufficiently descriptive" and "don't capture all nuances" are conflicting and a clear indication that they do need documentation. I understand the risk of them going out of sync and I'm opposed to remove them after refactor given that the code is sufficiently self descriptive, which is, IMO, at the current stage not the case.

Reply:
Discussed: keep for now, may be removed after refactor.